Add a test to show that the failure is not minimal

TheLortex opened the pull request.

talex5 commented:
In general, there is no one explanation for why a solve failed. This example is rather simple, as there's only one version of the package you want, and it depends on something that doesn't exist at all. I guess you could special-case that. But in general there will be many not-quite-solutions and it's useful to see a bit more context.

The example is also a bit confusing in that it lists e.g. ppxlib `0.1.0` above `0.3.1`, which the real system wouldn't do (higher version numbers are normally preferred).

I'm sure there are some things that could be done to improve the diagnostics, though. Highlighting the `problem` cases more clearly would be good (they tend to get lost amongst the "not the latest version" selections).

Another possibility would be to try to improve the initial diagnostic solution by trying to force each "problem" case to solve in turn. e.g. if we said that `ocaml-base-compiler` MUST succeed then we would get a better solution (OCaml 4.12 instead of 4.11). We could keep doing that as long at the number of problem cases got smaller.

Finally, after a failed solve it should probably display a message pointing you at a web-page explaining how to debug it.

TheLortex commented:
Hello ! Thank you for your answer. I've simplified and updated the tests so that they pinpoint exactly what I find is the problem. 

> The example is also a bit confusing in that it lists e.g. ppxlib 0.1.0 above 0.3.1, which the real system wouldn't do (higher version numbers are normally preferred).

Taken into account in the updated example.

> This example is rather simple, as there's only one version of the package you want, and it depends on something that doesn't exist at all.

I tried to simplify to what I thought was the "core" of the issue. Because my initial problem happens when solving with >50 dependencies so it would have been hard to reason on it. Now the "depends on something that doesn't exist at all" part is actually quite common in `opam-monorepo`, as we exclude all packages that doesn't depend on `dune`. 

Do you have ideas of situations where a "best-effort" mechanism (improving the failure to have a maximal set of solved packages)  would be harmful to the diagnostics ?

talex5 commented:
> Do you have ideas of situations where a "best-effort" mechanism (improving the failure to have a maximal set of solved packages) would be harmful to the diagnostics ?

The way I'd expect it to fail would be by selecting an extremely old version of something (though that can happen anyway). The current system is biased towards problems showing recent versions. Choosing what to display as the example problem is always a heuristic and will always be wrong sometimes, but I think it would be worth trying to improve the failure examples that way.

opam uses a different approach I think, where it tries to explain more generally why no solution was possible. But I find that output even harder to understand. The 0install way is to show a concrete example of a problem and then have the user explore it interactively by pinning down particular versions where they think the example shown is wrong.
